File: 30_ENGINE/pvm/signals.py
# -*- coding: utf-8 -*-
"""
THE INPUT CONTRACT.

`AssetSignals` is everything the classifier is allowed to see. The rule that makes
this file worth having: **a field may exist here only if a real iOS device can
produce it for a real photo.** PhotoKit metadata, a Vision OCR string, Vision scene
labels, Vision face clusters, a hash we computed ourselves. Nothing else.

That constraint is the whole defence against a classifier that scores beautifully in
evaluation and cannot run on a phone. If a rule needs a signal that is not in this
struct, the honest answer is that the rule cannot be implemented yet — not that the
struct should quietly grow a field the device cannot fill.

Every signal carries what it cost to acquire (`Tier`), because the engine is required
to stop escalating the moment the answer stops changing (L1-B, Minimum Necessary
Inference). A classifier that always reads every signal is not cheaper to be right
with; it is just more expensive to be wrong with.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from datetime import datetime
from enum import IntEnum
from typing import List, Optional, Tuple


class Tier(IntEnum):
    """What a signal costs, in ascending order — and the engine escalates in exactly
    this order, stopping as soon as the answer stops changing.

    The ordering is not alphabetical or conceptual. OCR sits last because C-1 predicts
    it dominates per-asset cost by roughly an order of magnitude over the visual
    stages, which is why the harness gates it rather than running it universally. If
    the T0-A scale sweep contradicts that, this enum is what moves."""
    METADATA = 0   # free — already in the PhotoKit row
    HASH = 1       # one thumbnail decode plus a few hundred bytes of arithmetic
    VISUAL = 2     # scene labels / feature print
    FACES = 3      # face detection and cluster lookup
    TEXT = 4       # gated OCR — the expensive one C-1 exists to ration

    @property
    def label(self) -> str:
        return {0: "metadata", 1: "hash", 2: "visual", 3: "faces", 4: "ocr"}[int(self)]


@dataclass(frozen=True)
class GeoFix:
    """A GPS fix carried by the asset itself. §11: an inferred location must declare
    lower confidence than a measured one, and the two must never be conflated."""
    lat: float
    lon: float
    accuracy_m: float = 65.0
    source: str = "exif"          # "exif" (measured) | "inferred" (from neighbours)

    @property
    def confidence(self) -> float:
        return 1.0 if self.source == "exif" else 0.6

    def km_to(self, other: "GeoFix") -> float:
        r = 6371.0
        p1, p2 = math.radians(self.lat), math.radians(other.lat)
        dp = p2 - p1
        dl = math.radians(other.lon - self.lon)
        a = math.sin(dp / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dl / 2) ** 2
        return 2 * r * math.asin(min(1.0, math.sqrt(a)))


@dataclass(frozen=True)
class PlaceName:
    """Reverse-geocoded names. Separate from GeoFix because geocoding can fail while
    the coordinates remain perfectly good."""
    country: Optional[str] = None
    city: Optional[str] = None
    confidence: float = 0.0


@dataclass(frozen=True)
class SceneLabel:
    identifier: str
    confidence: float


@dataclass(frozen=True)
class FaceCluster:
    """Vision gives a cluster; the user gives it a name. An unnamed cluster is still
    evidence that a person is present — that alone changes the risk class."""
    cluster_id: str
    name: Optional[str] = None
    area_fraction: float = 0.0


@dataclass
class AssetSignals:
    asset_id: str

    # ---- Tier.METADATA -----------------------------------------------------
    created_at: Optional[datetime] = None
    modified_at: Optional[datetime] = None
    pixel_w: int = 0
    pixel_h: int = 0
    byte_size: int = 0
    media_type: str = "image"              # "image" | "video"
    duration_s: float = 0.0
    is_screenshot: bool = False
    is_screen_recording: bool = False
    is_panorama: bool = False
    is_live_photo: bool = False
    # Where the pixels came from. On device this is derived from PHAssetResource
    # and the absence of camera EXIF, not guessed from the filename alone.
    source: str = "unknown"                # camera | screenshot | downloaded | shared | unknown
    burst_id: Optional[str] = None
    filename: Optional[str] = None
    geo: Optional[GeoFix] = None
    place: Optional[PlaceName] = None

    # ---- Tier.HASH ---------------------------------------------------------
    content_hash: Optional[str] = None     # exact bytes
    dhash: Optional[int] = None            # 64-bit perceptual

    # ---- Tier.TEXT ---------------------------------------------------------
    ocr_ran: bool = False
    ocr_text: str = ""

    # ---- Tier.VISUAL -------------------------------------------------------
    scene_labels: List[SceneLabel] = field(default_factory=list)
    embedding: Optional[bytes] = None

    # ---- Tier.FACES --------------------------------------------------------
    face_clusters: List[FaceCluster] = field(default_factory=list)

    # ---- derived, free -----------------------------------------------------
    @property
    def aspect_ratio(self) -> float:
        return (self.pixel_h / self.pixel_w) if self.pixel_w else 0.0

    @property
    def megapixels(self) -> float:
        return self.pixel_w * self.pixel_h / 1e6

    @property
    def year(self) -> Optional[int]:
        return self.created_at.year if self.created_at else None

    @property
    def named_people(self) -> List[str]:
        return sorted({f.name for f in self.face_clusters if f.name})

    @property
    def has_usable_dhash(self) -> bool:
        """FC-1a, and this is not pedantry: `dHash` returns 0 both for a whole class
        of ordinary images and for every hash failure. Treating 0 as a value would
        link unrelated photos as duplicates and make a failed hash indistinguishable
        from a confident match. Zero is therefore not a hash here — it is a miss."""
        return self.dhash is not None and self.dhash != 0


def hamming(a: int, b: int) -> int:
    return bin(a ^ b).count("1")

File: 40_APP/PVMCore/Sources/PVMCore/Signals.swift
import Foundation

/// THE INPUT CONTRACT — everything the classifier is allowed to see.
///
/// A field may exist here only if a real iPhone can produce it for a real photo:
/// PhotoKit metadata, a Vision OCR string, Vision scene labels, Vision face clusters,
/// a hash we computed ourselves. Nothing else. That constraint is what stops a
/// classifier that scores beautifully in evaluation and cannot run on a phone.
///
/// Mirrors `30_ENGINE/pvm/signals.py`. The engine is the reference implementation;
/// this is what ships.

/// What a signal costs, ascending — and the order the engine escalates in.
/// OCR sits last because C-1 predicts it dominates per-asset cost. The T0-A scale
/// sweep measured the embedding dominating instead, on a host with no Neural Engine;
/// that finding is IN DOUBT rather than settled, so the order stands until a device
/// says otherwise.
public enum Tier: Int, Comparable, CaseIterable {
    case metadata = 0
    case hash = 1
    case visual = 2
    case faces = 3
    case text = 4

    public var label: String {
        switch self {
        case .metadata: return "metadata"
        case .hash: return "hash"
        case .visual: return "visual"
        case .faces: return "faces"
        case .text: return "ocr"
        }
    }

    public static func < (a: Tier, b: Tier) -> Bool { a.rawValue < b.rawValue }
}

/// A GPS fix carried by the asset itself. §11: an inferred location must declare lower
/// confidence than a measured one, and the two must never be conflated.
public struct GeoFix {
    public enum Source: String { case exif, inferred }

    public let lat: Double
    public let lon: Double
    public let source: Source

    public init(lat: Double, lon: Double, source: Source = .exif) {
        self.lat = lat
        self.lon = lon
        self.source = source
    }

    public var confidence: Double { source == .exif ? 1.0 : 0.6 }

    public func km(to other: GeoFix) -> Double {
        let r = 6371.0
        let p1 = lat * .pi / 180, p2 = other.lat * .pi / 180
        let dp = p2 - p1
        let dl = (other.lon - lon) * .pi / 180
        let a = sin(dp / 2) * sin(dp / 2) + cos(p1) * cos(p2) * sin(dl / 2) * sin(dl / 2)
        return 2 * r * asin(min(1.0, sqrt(a)))
    }
}

/// Reverse-geocoded names, separate from the fix because geocoding can fail while the
/// coordinates stay perfectly good.
public struct PlaceName {
    public let country: String?
    public let city: String?
    public let confidence: Double

    public init(country: String?, city: String?, confidence: Double) {
        self.country = country
        self.city = city
        self.confidence = confidence
    }
}

public struct SceneLabel {
    public let identifier: String
    public let confidence: Double
    public init(identifier: String, confidence: Double) {
        self.identifier = identifier
        self.confidence = confidence
    }
}

/// Vision gives a cluster; the user gives it a name. An unnamed cluster is still
/// evidence that a person is present, and that alone changes the risk class.
public struct FaceCluster {
    public let clusterID: String
    public let name: String?
    public let areaFraction: Double
    public init(clusterID: String, name: String? = nil, areaFraction: Double = 0) {
        self.clusterID = clusterID
        self.name = name
        self.areaFraction = areaFraction
    }
}

public struct AssetSignals {
    public var assetID: String

    // Tier.metadata — free, already in the PhotoKit row
    public var createdAt: Date?
    public var modifiedAt: Date?
    public var pixelW: Int = 0
    public var pixelH: Int = 0
    public var byteSize: Int64 = 0
    public var isVideo = false
    public var durationSeconds: Double = 0
    public var isScreenshot = false
    public var isScreenRecording = false
    public var burstID: String?
    /// Where the pixels came from — derived from PHAssetResource and the absence of
    /// camera EXIF, never guessed from a filename alone.
    public var source: String = "unknown"       // camera | screenshot | downloaded | shared | unknown
    public var geo: GeoFix?
    public var place: PlaceName?

    // Tier.hash
    public var contentHash: String?
    public var dhash: UInt64?

    // Tier.text
    public var ocrRan = false
    public var ocrText: String = ""

    // Tier.visual / .faces
    public var sceneLabels: [SceneLabel] = []
    public var faceClusters: [FaceCluster] = []

    public init(assetID: String) { self.assetID = assetID }

    public var aspectRatio: Double { pixelW > 0 ? Double(pixelH) / Double(pixelW) : 0 }

    public var namedPeople: [String] {
        Array(Set(faceClusters.compactMap { $0.name })).sorted()
    }

    /// FC-1a, and this is not pedantry: `dHash` returns 0 both for a whole class of
    /// ordinary images and for every hash failure. Treating 0 as a value would link
    /// unrelated photos as duplicates and make a failed hash indistinguishable from a
    /// confident match. Zero is a miss, not a hash.
    public var hasUsableDHash: Bool { (dhash ?? 0) != 0 }

    public var year: Int? {
        guard let d = createdAt else { return nil }
        return Calendar(identifier: .gregorian).component(.year, from: d)
    }
}

public func hamming(_ a: UInt64, _ b: UInt64) -> Int { (a ^ b).nonzeroBitCount }
